source/client_channel_handler.c

```diff
@@ -820,34 +775,58 @@ uint16_t mqtt_create_request(
     aws_mqtt_send_request_fn *send_request,
     void *send_request_ud,
     aws_mqtt_op_complete_fn *on_complete,
-    void *on_complete_ud) {
+    void *on_complete_ud,
+    bool noRetry) {
 
     AWS_ASSERT(connection);
     AWS_ASSERT(send_request);
-    struct aws_mqtt_outstanding_request *next_request = NULL;
+    struct aws_mqtt_request *next_request = NULL;
     { /* BEGIN CRITICAL SECTION */
         mqtt_connection_lock_synced_data(connection);
 
         if (connection->synced_data.state == AWS_MQTT_CLIENT_STATE_DISCONNECTING) {
             mqtt_connection_unlock_synced_data(connection);
-            /* User requested disconnecting, ensure no new requests are made until the channel finished shutting down.
-             */
+            /* User requested disconnecting, ensure no new requests are made until the channel finished shutting
+             * down. */
             AWS_LOGF_ERROR(
-                AWS_LS_MQTT_CLIENT, "id=%p: Disconnect requested, stop creating any new request.", (void *)connection);
+                AWS_LS_MQTT_CLIENT,
+                "id=%p: Disconnect requested, stop creating any new request until disconnect process finishes.",
+                (void *)connection);
             aws_raise_error(AWS_ERROR_MQTT_CONNECTION_DISCONNECTING);
             return 0;
         }
 
+        if (noRetry && connection->synced_data.state != AWS_MQTT_CLIENT_STATE_CONNECTED) {
```

Comment:
Options:
1. invoke complete callback & clean it up during the task. Problem is the channel may not exist, it will still be a synced call.
2. Fail the call. Exception with no offline queueing for QoS 0.
3. For ping, it's not an error and no complete callback, it's fine to do this. But, for publish, this will not invoke the on_complete callback, it should be consider as error?

Reply:
don't invoke completion callback synchonously